src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java

```diff
@@ -597,20 +579,24 @@ public static boolean hasChannel(ICommonPacketListener listener, ResourceLocatio
      * @return True if the connection has a connection setup that can transmit the given payload id, false otherwise.
      */
     public static boolean hasChannel(Connection connection, @Nullable ConnectionProtocol protocol, ResourceLocation payloadId) {
-        NetworkPayloadSetup payloadSetup = connection.channel().attr(ATTRIBUTE_PAYLOAD_SETUP).get();
-        if (payloadSetup == null) {
-            return getKnownAdHocChannelsOfOtherEnd(connection).contains(payloadId);
+        NetworkPayloadSetup payloadSetup = ChannelAttributes.getPayloadSetup(connection);
+        if (payloadSetup != null) {
+            // If a protocol is specified, only check against channels for that protocol
+            // Otherwise check against all protocols.
+            if (protocol != null && payloadSetup.getChannels(protocol).containsKey(payloadId)) {
+                return true;
+            } else if (protocol == null && payloadSetup.channels().values().stream().anyMatch(map -> map.containsKey(payloadId))) {
```

Comment:
Just a small fun comment: it took me looking at the code to realize why this looked strange to me -- `NetworkPayloadSetup` contains a map whose values are also maps. 😆